### DISCUSSION TITLE
How do you go about responsiveness?

### DISCUSSION THREAD

**Discussion by pettomartino** [vercel/next.js#13356] [Help]
Hello guys, I'm sorry if this is not the right place, or if it's too stupid question, but I would like to understand better how to go about responsive design in nextjs/server side rendering.

My example will use `react-responsive` library (https://github.com/contra/react-responsive) but the idea is very simple: if screen is bigger than X display A else display B.

So, let's say I have this component:

```jsx
import { useMediaQuery } from 'react-responsive'
export default function Home() {

  const isDesktopOrLaptop = useMediaQuery({
    query: '(min-device-width: 1224px)'
  })

  return (
    <div>
      {isDesktopOrLaptop && <div>
        This is desktop
      </div>}

      {!isDesktopOrLaptop && <div>
        This is mobile
      </div>}
    </div>
  )
}
```

This is pretty simple NextJS page that should show "This is desktop" if `width>= 1224px` or "This is mobile" if not.

However, I get this warning
![image](https://user-images.githubusercontent.com/101433/82829135-71da9200-9eaa-11ea-85b1-5b480bea5325.png)

I would like to know if there's a way to use this kind of strategy rather than CSS `display: none` because I want to prevent react rendering unnecessary components.
So let's say I have "deep" difference between those 2, would be better to prevent react from rendering it, right?

Thanks in advance!

**Comment by mayankguptadotcom** [comment]
@pettomartino Not sure what you feel about using Tailwindcss with NextJS.

**Reply by pettomartino** [reply]
@mayankguptadotcom Sorry, but what does it has to do with my question? Is it because tailwindcss has responsive css? My question is about preventing unnecessary rendering, css doesn't help with that.

**Reply by cdock1029** [reply]
I think using css media queries all throughout the react tree could be the most efficient though, as you're letting the browser do what it is good at and not branching within javascript/react trees but using css where needed to do so. The "DOM" tree, as much as possible, would remain the same, and css would determine differences. Though react might "re-render", the DOM won't be changing so much.

This article covers some strategies, though https://nitayneeman.com/posts/combining-server-side-rendering-and-responsive-design-using-react/

`getServerSideProps` would be useful for you to branch on user agent and have access to this at server render time. But it isn't completely dependable and also you miss out on static optimizations having to render on server for every request, instead of pre-render/serve static from CDN, better performance etc.

But from a bigger picture perspective, worrying too much about react rendering things might be an un-necessary optimization. Executing the render function doesn't have to be expensive. In fact you will find React itself often executes the render function mulitple times for given component. The reconciler determines if actual DOM updates are necessary if virtual DOM has changed. Often they are not, and it's not something you have to worry about by default... only in specific performance issues.

Unless there are significant differences in device branches, and if the large react tree has many side effects, that could be undesirable to render incorrect version. Only other way would be to render nothing (`null`) or some minimal common UI, then after client hydration in `useEffect` update state to render correct branch. But not server rendering something then hydrating a different branch (as your error shows which could introduce hydration bugs).

To sum up long-winded reply, if you really need it, I think you have to defer rendering decision till client hydration and use your hook (which needs browser apis to branch) and dont waste render on server (since you don't know what you need to render yet), render null. Your initial render wont show anything (perhaps loading or placeholder/shell) where these branches exist. Extract out common parts, use css where possible (trees that dont have different side effect operations, just styles).

**Comment by lenis0012** [comment]
I know this an old dicussion but I am running into the same issue.
I checked some major websites to see what they do. notably AirBnB (they dont use next, i think) seems to render a completely different div for mobile then they do for desktop if you look at for instance the navbar.

I wanted to replicate this behavior. but it is not possible server side. I read the article that @cdock1029 mentioned as well as some others. And using the user agent looks like trouble. you could just use the user agent to detect android/iOS and then render with a width of 500px passed on to responsive-react. but it wont always be accurate and might cause the app to re-render large components upon hydration.

without SSR/CSR using javascript media queries looks like the best solution. and thats what AirBnB does I think (though their classes are obfuscated so im not 100% sure)

That being said, with nextjs SSR/CSR I feel like css media queries are a better option.
I think it would be possible to render both versions in SSR, and only using react-responsive media queries in client-side hydration, but that optimization might not even be worth the trouble as it will have no effect on first meaningful paint.

I would be interested to hear if anyone is using any of the above, and how that works out for you.
The end goal of course, is to optimize your SEO score.

**Comment by redbar0n** [comment]
@lenis0012 Check out: [Fresnel](https://github.com/artsy/fresnel)

>An SSR compatible approach to CSS media query based responsive layouts for React.

It does not rely on `user-agent` but renders all breakpoints server-side. You could even combine it with [detect-responsive-traits](https://github.com/artsy/detect-responsive-traits) to optimise it to only render specific breakpoints if `user-agent` info is reliably determined. NB: It currently has some compatibility issues with React Native Web (which presumes no usage of media queries).

**Reply by redbar0n** [reply]
If you are using React Native Web, you might be interested in this breakdown of the current options: https://github.com/nandorojo/dripsy/discussions/109#discussioncomment-848718

**Comment by aayushmau5** [comment]
I faced the same situation quite some time ago, and the only solution I have found is to render the UI on the client side using `useEffect` hook.
```js
import { useEffect, useState } from "react";
import { useMediaQuery } from "react-responsive";

export default function Nav() {
  const [mounted, setMounted] = useState(false);

  useEffect(() => {
    setMounted(true);
  }, []);

  return (
    <>
      {mounted && (
        <div>
          <Desktop>Desktop or laptop</Desktop>
        </div>
      )}
    </>
  );
}

function Desktop({children}) {
  const useDesktopMediaQuery = () =>
    useMediaQuery({
      minWidth: 769,
    });

 return (
  <>
   { useDesktopMediaQuery() ? children : null; };
  </>
);
}
```
As `useEffect` will only run when the component has mounted, the whole UI will render on the client-side. It's always not a bad idea to render the UI on client-side(unless you have really big and heavy UI).

**Reply by redbar0n** [reply]
Thanks for sharing your solution. So the hydrated content will match SSR’ed content on the first render client-side. But on second render, when ‘mounted’ is true, won’t it potentially introduce flicker/jank if/when the markup is different (and a different media query applies) than what was first rendered?

**Reply by aayushmau5** [reply]
Yeah, you are right. I may introduce some flicker if the markup is different. I never gave that a thought until you pointed it out because I used it for a responsive navigation bar and the markup is not that different. :slightly_smiling_face: 

For that matter, what do you recommend? I'm looking at the library you provided in answer above. Will that solve the "flicker" problem?

Thanks :slightly_smiling_face:

**Reply by redbar0n** [reply]
@aayushmau5 Thanks for confirming. For CSR (no NextJS) you could use your solution, or avoid media queries altogether and just get window dimensions in JS ([useWindowDimensions](https://necolas.github.io/react-native-web/docs/use-window-dimensions/) in RN/RNW). For SSR, media queries is best.

[Fresnel](https://github.com/artsy/fresnel), potentially also with [detect-responsive-traits](https://github.com/artsy/detect-responsive-traits), can work. But both are optimisations that complicates things a bit.

In general, I would try to avoid "breaking web presumptions" (e.g. 'one URL leads to one HTML document', and 'CSS media queries are for alternating CSS') by rendering and sending potentially wholly different markup to the client. Rather, I would only use media queries for styling only, and always SSR and send a single markup tree (one HTML document) to the client. That way you know React will be able to hydrate and render correctly on the client, without artefacts. The compromise is that you have to choose what markup to send, once and for all: desktop, tablet or mobile. **In general, esp. if using RNW, I would just choose mobile-first design/rendering. It's most straightforward.** Your PWA can then be SSR'ed, without client-side artefacts appearing later.

> **Mobile-first:** SSR only mobile page markup, with pre-determined/static breakpoints (no Fresnel). Let the end-user choose to "Show desktop version" on first page, then set cookie for all future requests. Avoids rendering multiple layouts/breakpoints/sub-trees (like with vanilla Fresnel), so less CSS & HTML is sent over-the-wire (again optimising for mobile clients).

[Others](https://github.com/nandorojo/dripsy/discussions/109) have chosen to sacrifice SSR and media queries in favor of CSR with JS, to be able to have responsivity that also changes the markup rendered to fit any client type (like in your example). They additionally use NextJS with SSG to get SEO. So it's not all black–and-white, and this might be a good option if your page is not that dynamic, with a finite amount of pages.

You can see the discussion here, where I go into a bit more detail on alternatives: [How to combine React Native Web (RNW) + Responsive styles (media queries) + NextJS Server-Side Rendering (SSR), to get SEO on the web](https://github.com/nandorojo/dripsy/discussions/109#discussioncomment-848718).

**Reply by aayushmau5** [reply]
Wow, that was pretty good and detailed explanation. Thanks a lot for clarifying it :slightly_smiling_face:

**Reply by redbar0n** [reply]
My pleasure :-) Oh, btw, you might not even need media queries to get responsive components: 

>Do you really need a media query?
Flexbox, Grid, and multi-column layout all give you ways to create flexible and even responsive components without the need for a media query. 

https://developer.mozilla.org/en-US/docs/Learn/CSS/CSS_layout/Media_queries#do_you_really_need_a_media_query

**Reply by iamyoki** [reply]
I wouldn't use `useEffect` to calculate something before rendering.
```js
 const [mounted, setMounted] = useState(false);

  useEffect(() => {
    setMounted(true);
  }, []);

  return (
    <>
      {mounted && (
        <div>
          <Desktop>Desktop or laptop</Desktop>
        </div>
      )}
    </>
  );
```

Instead, `useLayoutEffect` was meant to achieve that.
```js
const [isMobile, setIsMobile] = useState(false)
useLayoutEffect(()=>{
  ....
  setIsMobile(true)
}, [])

return isMobile ? <MobileApp /> : <App />
```

**Reply by rubek-joshi** [reply]
> Instead, `useLayoutEffect` was meant to achieve that.

Thing is you can't use `useLayoutEffect` with SSRs, SSGs and ISRs since the window is `undefined`.

**Comment by hhimanshu** [comment]
I solved this problem like below

```ts
import MobileHome from './MobileHome'
import DesktopHome from './DesktopHome'
import dynamic from "next/dynamic"

const MediaQuery = dynamic(() => import("react-responsive"), {
    ssr: false
})
export default function Home() {
    return <>
        <MediaQuery maxWidth={767}><MobileHome/></MediaQuery>
        <MediaQuery minWidth={768}><DesktopHome/></MediaQuery>
    </>
}
```

**Comment by theonlydaleking** [comment]
[Sizes](https://nextjs.org/docs/pages/api-reference/components/image#sizes) is probably what you're after

**Comment by monolithed** [comment]


**Reply by theonlydaleking** [reply]
-> Telling NextJS to remove SSR 😂

**Comment by gerardnico** [comment]
You can also check the user agent and redirect the request to a dedicated page.

Example:

* Request Mobile Site - "User-Agent": "Mozilla/5.0 (iPhone; CPU iPhone OS 16_1 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) CriOS/108.0.5359.52 Mobile/15E148 Safari/604.1"

* Request Desktop Site - "User-Agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_13_5) AppleWebKit/605.1.15 (KHTML, like Gecko) CriOS/108 Version/11.1.1 Safari/605.1.15"

**Reply by gerardnico** [reply]
In a server component, this is the only option. 

* You can also load a dummy page that do a redirect to the proper page once it loads if you are not sure of the device type / screen
* You can also take a bootstrap css approach by showing and hiding component with a css media query.

**Reply by redbar0n** [reply]
Fresnel can do the user agent thing for you. Some other approaches to solving the same problem with combining responsiveness / media queries and SSR: https://dev.to/redbar0n/how-to-combine-react-native-web-responsivity-nextjs-ssr-to-get-seo-a4g

**Comment by aravindanve** [comment]
I wanted to pre-render a page conditionally based on the device type on React/NextJS/SSR. I was able to do it by storing the user-agent in a context and reading it from a hook `useIsMobile()` which returned true or false based on the parsed user-agent header on the server side.

We create a context to store the parsed user-agent header.
```ts
// userAgentContext.ts
import { userAgent } from "next/server";
import { createContext } from "react";

export type UserAgent = ReturnType<typeof userAgent>;
export const UserAgentContext = createContext<UserAgent | undefined>(undefined);
```

We create a hook to consume the context and check device type.
```ts
// useIsMobile.ts
import { useContext } from "react";
import { UserAgentContext } from "./userAgentContext.ts";

export const useIsMobile = () => {
  const userAgent = useContext(UserAgentContext);
  if (!userAgent) {
    throw new Error("useIsMobile must be used within a UserAgentContext.Provider");
  }

  return userAgent.device.type === "mobile";
};
```

We can use a sever component as the layout (app router) to get the user agent from request headers
```tsx
// layout.tsx
"use server";

import { headers } from "next/headers";
import { userAgent } from "next/server";
import { LayoutClient } from "./layoutClient.ts";

export default async function Layout({ children }: { children: React.ReactNode }) {
  const reqUserAgent = userAgent({ headers: headers() });
  return <LayoutClient reqUserAgent={reqUserAgent}>{children}</LayoutClient>
}

```

We can then pass the user agent to a client component that will provide the initial value for the context. Since client components with `"use client"` directive are still rendered once on the server with initial values then hydrated on the client.
```tsx
// layoutClient.tsx
"use client";

import { UserAgentContext } from "./userAgentContext.ts";

export default function LayoutClient({ reqUserAgent, children }: { reqUserAgent: UserAgent, children: React.ReactNode }) {
  return <UserAgentContext.Provider value={reqUserAgent}>{children}</UserAgentContext.Provider>;
}

```

Now when this page is rendered for the first time on the server with initial values, useIsMobile() should accurately return a value based on the user-agent device type.
```tsx
// page.tsx
"use client";

export default function Page() {
  const isMobile = useIsMobile();

  return isMobile ? <>Mobile</> : <>Desktop</>;
}
```

**Reply by monolithed** [reply]
`isMobile` is a magical variable that doesn't provide any information about screen proportions and image rendering. As a designer, how would you draw a layout for iPad, iPhone X, iPhone XS, iPhone 15 Pro Max, and so on, based solely on this variable?

**Reply by aravindanve** [reply]
@monolithed This is just an example. You can rewrite the hook to suit your needs, you can even return the whole userAgent object if you want to decide based on browser, device and other informatrion.

```ts
export const useUserAgent = () => {
  const userAgent = useContext(UserAgentContext);
  if (!userAgent) {
    throw new Error("useUserAgent must be used within a UserAgentContext.Provider");
  }

  return userAgent;
};
```

```ts
const userAgent = useUserAgent();

return userAgent.browser.name === "chrome" ? ...
```

Or you can pass userAgent to a library like `react-device-detect`:

```tsx
const userAgent = useUserAgent();
const { isIOS, isIPad13 } = getSelectorsByUserAgent(userAgent);

return isIPad13 ? <>IPad 13</> : isIOS ? <>IOS</> : <>Other</>;
```

**Reply by monolithed** [reply]
Modern product interfaces are created not for specific devices, but for screen sizes. Not only are there thousands of devices, but browsers often change their user agents, and in some cases even present themselves as other manufacturers. Relying on user agents is justified only in one case - when you need to provide the user with a binary file for download. However, even in this scenario, software manufacturers are moving away from it because supporting user agents requires significant resources. Relying on the userAgent takes us back to the 90s.

**Reply by aravindanve** [reply]
@monolithed I think you're missing the point as to what this is for. 

You should ideally use CSS media queries for responsiveness wherever you can. But there are scenarios where you'd need to programmatically render different menus or use a slightly different layout based on screen size, and a pure CSS solution is extremely difficult. In such cases server rendering a page will cause it to always pre-render for the default device. I.e. this sort of behaviour is primarily achieved through hooks like antd's useBreakpoint, but it does not work on the server. So if your page is mobile first, then the server always renders a mobile page, then it flashes on your desktop browser for a second before useBreakpoint kicks in and it changes to the desktop site. 

The solution is to use something like what I've suggested above when you have no other data regarding the device while server rendering a page. It is not a generic solution to all your responsiveness issues. 

In my case I ended up wrapping useBreakpoint in my own hook useServerBreakpoint which returns false for xl, lg, md if the user agent is mobile. It's not perfect but I'm not sure a better solution exists at the moment.

**Reply by redbar0n** [reply]
Here are some other approaches on how to solve responsiveness / media queries with SSR: https://dev.to/redbar0n/how-to-combine-react-native-web-responsivity-nextjs-ssr-to-get-seo-a4g

**Reply by jsefiani** [reply]
@monolithed It is not ideal, but another viable solution doesn't seem to exist. Solutions like react-responsive or building your own media query hook don't work in server land because there is no window object available, so you will always get some layout flickering on the first render/reload, which I don't want for my users.

Using the user-agent should be reliable until we get a better solution. 

I am pretty sure other companies use a similar approach for adaptive design, which is not the same as responsive design.

The latter is much easier to achieve using just media queries.

@aravindanve, thanks for the solution!

PS: One potential improvement might be adding a resize event handler when a user flips their mobile device.
```
export const useIsMobile = () => {
  const userAgent = useContext(UserAgentContext);
  const [isMobile, setIsMobile] = useState(userAgent?.device.type === "mobile");

  if (!userAgent) {
    throw new Error(
      "useIsMobile must be used within a UserAgentContext.Provider"
    );
  }

  useEffect(() => {
    const checkMobile = () => setIsMobile(window.innerWidth <= 768);

    window.addEventListener("resize", checkMobile);
    return () => window.removeEventListener("resize", checkMobile);
  }, []);

  return isMobile;
};
```

**Reply by aravindanve** [reply]
@jsefiani Yes, and if you use **antd** like I do, you can just wrap their `Grid.useBreakpoint` hook. It automatically takes care of window resize and all breakpoints, you just need to handle the default values on the server:

```ts
/**
 * Estimates breakpoints on server side based on device detection to prevent
 * always loading mobile layout on SSR
 **/
export const useServerBreakpoint = () => {
  const breakpoint = Grid.useBreakpoint();
  const userAgent = useContext(UserAgentContext);

  if (!userAgent) {
    throw new Error("useServerBreakpoint must be used within a UserAgentProvider");
  }

  const isMobile = userAgent.device.type === "mobile";

  return useMemo(
    () => ({
      xxl: breakpoint.xxl ?? !isMobile,
      xl: breakpoint.xl ?? !isMobile,
      lg: breakpoint.lg ?? !isMobile,
      md: breakpoint.md ?? !isMobile,
      sm: breakpoint.sm ?? true,
      xs: breakpoint.xs ?? true,
    }),
    [breakpoint.lg, breakpoint.md, breakpoint.sm, breakpoint.xl, breakpoint.xs, breakpoint.xxl, isMobile],
  );
};
```
